File: typescript/test/serialize.test.ts
import { describe, it, expect } from "vitest";
import type { Shape, Vehicle } from "./testTypes.js";

// Interop JSON — identical strings used across all 12 language libraries
const CIRCLE_JSON = '{"#type":"Circle","radius":4}';
const RECTANGLE_JSON = '{"#type":"Rectangle","width":10,"height":5}';
const CAR_JSON = '{"#type":"Car","make":"Toyota"}';
const TRUCK_JSON = '{"#type":"Truck","payloadTons":5}';

describe("JSON.stringify", () => {
  it("when serializing Circle, matches interop format", () => {
    // Arrange
    const shape: Shape = { "#type": "Circle", radius: 4 };

    // Act
    const json = JSON.stringify(shape);

    // Assert
    expect(json).toBe(CIRCLE_JSON);
  });

  it("when serializing Rectangle, matches interop format", () => {
    // Arrange
    const shape: Shape = { "#type": "Rectangle", width: 10, height: 5 };

    // Act
    const json = JSON.stringify(shape);

    // Assert
    expect(json).toBe(RECTANGLE_JSON);
  });

  it("when serializing Car with custom tag name, matches interop format", () => {
    // Arrange
    const vehicle: Vehicle = { "#type": "Car", make: "Toyota" };

    // Act
    const json = JSON.stringify(vehicle);

    // Assert
    expect(json).toBe(CAR_JSON);
  });

  it("when serializing Truck with custom tag name, matches interop format", () => {
    // Arrange
    const vehicle: Vehicle = { "#type": "Truck", payloadTons: 5 };

    // Act
    const json = JSON.stringify(vehicle);

    // Assert
    expect(json).toBe(TRUCK_JSON);
  });
});

File: ocaml/test/serialize_test.ml
(* Interop JSON — identical strings used across all 12 language libraries *)
let circle_json = {|{"#type":"Circle","radius":4}|}
let rectangle_json = {|{"#type":"Rectangle","width":10,"height":5}|}
let car_json = {|{"#type":"Car","make":"Toyota"}|}
let truck_json = {|{"#type":"Truck","payloadTons":5}|}

(* Yojson emits .0 for float values *)
let yojson_circle_json = {|{"#type":"Circle","radius":4.0}|}
let yojson_rectangle_json = {|{"#type":"Rectangle","width":10.0,"height":5.0}|}
let yojson_truck_json = {|{"#type":"Truck","payloadTons":5.0}|}

let tests =
  [
    Alcotest.test_case
      "encode, when serializing circle, matches interop format" `Quick
      (fun () ->
        (* Act *)
        let json =
          Test_types.encode_shape Json_tag.encode (Test_types.Circle { radius = 4.0 })
          |> Yojson.Safe.to_string
        in
        (* Assert *)
        Alcotest.(check string) "circle json" yojson_circle_json json);
    Alcotest.test_case
      "encode, when serializing rectangle, matches interop format" `Quick
      (fun () ->
        (* Act *)
        let json =
          Test_types.encode_shape Json_tag.encode (Test_types.Rectangle { width = 10.0; height = 5.0 })
          |> Yojson.Safe.to_string
        in
        (* Assert *)
        Alcotest.(check string) "rectangle json" yojson_rectangle_json json);
    Alcotest.test_case
      "encode, when serializing car, matches interop format" `Quick
      (fun () ->
        (* Act *)
        let json =
          Test_types.encode_vehicle Json_tag.encode (Test_types.PersonalVehicle { make = "Toyota" })
          |> Yojson.Safe.to_string
        in
        (* Assert *)
        Alcotest.(check string) "car json" car_json json);
    Alcotest.test_case
      "encode, when serializing truck, matches interop format" `Quick
      (fun () ->
        (* Act *)
        let json =
          Test_types.encode_vehicle Json_tag.encode (Test_types.TransportationVehicle { payload_tons = 5.0 })
          |> Yojson.Safe.to_string
        in
        (* Assert *)
        Alcotest.(check string) "truck json" yojson_truck_json json);
  ]
